compiler/noirc_evaluator/src/ssa/opt/fill_internal_slices.rs

```diff
@@ -0,0 +1,397 @@
+use std::collections::BTreeMap;
+
+use crate::ssa::{
+    ir::{
+        basic_block::BasicBlockId,
+        function::Function,
+        function_inserter::FunctionInserter,
+        instruction::Instruction,
+        post_order::PostOrder,
+        types::Type,
+        value::{Value, ValueId},
+    },
+    ssa_gen::Ssa,
+};
+
+use acvm::FieldElement;
+use fxhash::FxHashMap as HashMap;
+
+impl Ssa {
+    /// Fill out slice values represented by SSA array values to contain
+    /// dummy data that accounts for dynamic array operations in ACIR gen.
+    /// When working with nested slices in ACIR gen it is impossible to discern the size
+    /// of internal slices. Thus, we should use the max size of internal nested slices for reading from memory.
+    /// However, not increasing the capacity of smaller nested slices can lead to errors where
+    /// we end up reading past the specified dynamic index. If we want to also read information that goes
+    /// past the nested slice, we will have garbage data in its place if the nested structure is transformed to match.
+    pub(crate) fn fill_internal_slices(mut self) -> Ssa {
+        for function in self.functions.values_mut() {
+            let mut context = Context::new(function);
+            context.process_blocks();
+        }
+        self
+    }
+}
+
+struct Context<'f> {
+    post_order: PostOrder,
+    inserter: FunctionInserter<'f>,
+}
+
+impl<'f> Context<'f> {
+    fn new(function: &'f mut Function) -> Self {
+        let post_order = PostOrder::with_function(function);
+        let inserter = FunctionInserter::new(function);
+
+        Context { post_order, inserter }
+    }
+
+    fn process_blocks(&mut self) {
+        let mut block_order = PostOrder::with_function(self.inserter.function).into_vec();
+        block_order.reverse();
+        for block in block_order {
+            self.process_block(block);
+        }
+    }
+
+    fn process_block(&mut self, block: BasicBlockId) {
+        // Fetch SSA values potentially with internal slices
+        let instructions = self.inserter.function.dfg[block].take_instructions();
```

Comment:
Is this necessary? I don't see any changes to the dfg directly in the for loop below at least. If there are mutations within some of the functions that are called, maybe it'd be clearer to separate out those into a different step from the longest-length collection step.

Reply:
Yeah I will separate out the steps. The first loop is simply a collection step and only in the second loop do we go through and modify the relevant slices that need to be filled with dummy data by calling the `attach_slice_dummies` method and pushing a new instruction value.